src/inscription/iid.rs

```diff
@@ -0,0 +1,276 @@
+            .copied()
+            .collect()
+    }
+
+    /// Creates a new InscriptionId from a transaction's output reference.
+    pub fn from_outpoint(outpoint: OutPoint) -> Self {
+        Self {
+            txid: outpoint.txid,
+            index: outpoint.vout,
+        }
+    }
+
+    /// Creates a new InscriptionId from its string representation.
+    pub fn parse_from_str(iid: &str) -> OrdResult<Self> {
+        Self::from_str(iid).map_err(OrdError::InscriptionParser)
+    }
+}
+
+impl std::fmt::Display for InscriptionId {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}i{}", self.txid, self.index)
+    }
+}
+
+impl Serialize for InscriptionId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_str(self)
+    }
+}
+
+impl<'de> Deserialize<'de> for InscriptionId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        DeserializeFromStr::with(deserializer)
+    }
+}
+
+struct DeserializeFromStr<T: FromStr>(pub T);
+
+impl<'de, T: FromStr> DeserializeFromStr<T>
+where
+    T::Err: std::fmt::Display,
+{
+    pub fn with<D>(deserializer: D) -> Result<T, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(DeserializeFromStr::<T>::deserialize(deserializer)?.0)
+    }
+}
+
+impl<'de, T: FromStr> Deserialize<'de> for DeserializeFromStr<T>
+where
+    T::Err: std::fmt::Display,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Self(
+            FromStr::from_str(&String::deserialize(deserializer)?)
+                .map_err(serde::de::Error::custom)?,
+        ))
+    }
+}
+
+impl FromStr for InscriptionId {
+    type Err = InscriptionParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Some(char) = s.chars().find(|char| !char.is_ascii()) {
+            return Err(InscriptionParseError::Character(char));
+        }
+
+        const TXID_LEN: usize = 64;
+        const MIN_LEN: usize = TXID_LEN + 2;
+
+        if s.len() < MIN_LEN {
+            return Err(InscriptionParseError::InscriptionIdLength(s.len()));
+        }
+
+        let txid = &s[..TXID_LEN];
+
+        let separator = s.chars().nth(TXID_LEN).unwrap();
+
+        if separator != 'i' {
+            return Err(InscriptionParseError::CharacterSeparator(separator));
+        }
+
+        let vout = &s[TXID_LEN + 1..];
+
+        Ok(Self {
+            txid: txid.parse().map_err(InscriptionParseError::Txid)?,
+            index: vout.parse().map_err(InscriptionParseError::Index)?,
+        })
+    }
+}
+
+#[cfg(test)]
+fn set_using(n: u32) -> InscriptionId {
+    let hex = format!("{n:x}");
+
+    if hex.is_empty() || hex.len() > 1 {
+        panic!();
+    }
+
+    format!("{}i{n}", hex.repeat(64)).parse().unwrap()
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I think better to move `set_using` inside `mod tests{}`.

Reply:
Done